packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart

```diff
@@ -0,0 +1,117 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert' show json;
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+import 'package:shared_preferences_platform_interface/shared_preferences_platform_interface.dart';
+
+import 'src/win32.dart' as win32;
+
+/// Wrapper to the win32 ffi functions for testing.
+class Win32Wrapper {
+  /// Calls the win32 function for [getLocalDataPath].
+  String getLocalDataPath() {
+    return win32.getLocalDataPath();
+  }
+
+  /// Calls the win32 function for [getModuleFileName].
+  String getModuleFileName() {
+    return win32.getModuleFileName();
+  }
+}
+
+/// The Windows implementation of [SharedPreferencesStorePlatform].
+///
+/// This class implements the `package:shared_preferences` functionality for Windowds.
+class SharedPreferencesWindows extends SharedPreferencesStorePlatform {
+  /// The name of the parent directory for [fileName].
+  final String fileDirectory = 'Flutter';
+
+  /// File system used to store to disk. Exposed for testing only.
+  @visibleForTesting
+  FileSystem fs = LocalFileSystem();
+
+  /// Wrapper for win32 ffi calls.
+  @visibleForTesting
+  Win32Wrapper win32Wrapper = Win32Wrapper();
+
+  String _localDataFilePath;
+
+  /// The path to the file in disk were the preferences are stored.
+  @visibleForTesting
+  String get getLocalDataFilePath {
+    if (_localDataFilePath != null) {
+      return _localDataFilePath;
+    }
+    String appPath = win32Wrapper.getModuleFileName();
+    String appName = path.basenameWithoutExtension(appPath);
+    String localDataPath = win32Wrapper.getLocalDataPath();
+    // Append app-specific identifiers.
+    _localDataFilePath =
+        path.join(localDataPath, fileDirectory, '$appName.json');
+    return _localDataFilePath;
+  }
+
+  Map<String, Object> _cachedPreferences;
+
+  /// The in-memory representation of the map saved to a file in disk;
+  @visibleForTesting
+  Map<String, Object> get getCachedPreferences {
+    if (_cachedPreferences == null) {
+      _cachedPreferences = {};
+      File localDataFile = fs.file(getLocalDataFilePath);
+      if (localDataFile.existsSync()) {
+        String stringMap = localDataFile.readAsStringSync();
+        if (stringMap.isNotEmpty) {
+          _cachedPreferences = json.decode(stringMap) as Map<String, Object>;
+        }
+      }
+    }
+    return _cachedPreferences;
+  }
+
+  /// Writes the cached preferences to disk. Returns [true] if the operation
+  /// succeeded.
+  bool _writePreferencesToFile() {
+    try {
+      File localDataFile = fs.file(getLocalDataFilePath);
+      if (!localDataFile.existsSync()) {
+        localDataFile.createSync(recursive: true);
+      }
+      String stringMap = json.encode(getCachedPreferences);
+      localDataFile.writeAsStringSync(stringMap);
```

Comment:
Since all the callers are async, we should probably make this whole function async and change this to `await localDataFile.writeAsString(stringMap);`

Reply:
Same for reading, now that I think about it.

Reply:
What's the benefit on this? I would still have to call `await` on all the async functions. Then calls like `getCachedPreferences.clear();` would have to change to `await getCachedPreferences.clear();` which is less clean IMO.

Reply:
Discussed offline. There's no strong opinion on using `sync`. Will use the `*AsSync` methods.